### PR TITLE
feat: add BC pretraining device policy (#1213)

### LINKED ISSUE #1213: training: add config-driven BC pretraining device policy
## Summary
BC pretraining should expose an explicit config-driven `device` policy instead of staying hardcoded to CPU. The default should be `auto`, with `cpu` and accelerator-specific values available as explicit overrides.

## Decision
- Chosen policy: add an explicit BC pretraining `device` field.
- Default: `auto`.
- Rationale: BC warm-start training is expected to be too slow if the repository forces CPU-only execution. GPU acceleration should be used on capable hosts by default, while reproducible CPU runs remain available through an explicit `device: cpu` override.

## Why
- `pretrain_from_expert.py` currently hardcodes CPU for both the SB3 policy wrapper and the imitation BC trainer.
- The prior review feedback asking for `device="auto"` raised a real policy issue, not just a one-line code style concern.
- A config surface keeps runtime hardware selection visible in configs, docs, tests, and run provenance.

## Scope
In scope:
- Add a BC pretraining config/device surface with default `auto`.
- Thread the configured device into the SB3 policy wrapper and imitation BC trainer where applicable.
- Document how to force CPU for deterministic or resource-constrained runs.
- Add regression coverage for the selected default and override behavior.

Out of scope:
- Running the full BC warm-start PPO experiment.
- Changing PPO training device policy outside the BC pretraining stage.
- Introducing host-specific hardware assumptions beyond the existing `auto` behavior.

## Acceptance criteria
- `pretrain_from_expert.py` no longer hardcodes CPU as the only BC pretraining device policy.
- The default BC device policy is `auto`.
- A config or CLI path can explicitly force `cpu`.
- Docs or config comments explain when to use `auto` versus `cpu`.
- Tests cover default `auto` behavior and explicit CPU override plumbing without requiring a GPU.

## Validation
- Targeted tests for config parsing and device propagation.
- `uv run python scripts/dev/check_skills.py` if workflow docs or skills are touched.
- `BASE_REF=origin/main scripts/dev/pr_ready_check.sh` before PR handoff unless the implementation PR documents a narrower proof.

### PR DESCRIPTION
**Pull request by ll7**
## Summary
Adds an explicit config-driven BC pre-training `device` policy with `auto` as the default and CPU override support for deterministic or resource-constrained runs.

## Linked Issues
- Closes #1213

## What Changed
- Added `device` to `BCPretrainingConfig`, defaulting to `auto`.
- Threaded `config.device` into the Stable-Baselines3 PPO container and `imitation.algorithms.bc.BC` trainer.
- Added `device: auto` to the checked-in BC pre-training configs.
- Documented `auto` versus `cpu` in the imitation learning guide and dev guide.
- Added regression tests for default `auto`, explicit `cpu`, YAML override loading, and trainer propagation.

## Why It Matters
- Added value: BC warm-start hardware selection is now visible in config and reproducible in run handoff.
- Expected impact: GPU-capable hosts can use accelerator defaults without hardcoding CPU-only BC training; CPU remains an explicit opt-in.
- Why this is worth merging now: it resolves the policy ambiguity tracked in #1213 before the BC warm-start experiment path relies on hidden library behavior.

## Validation / Proof
- Commands run:
  - `uv run pytest tests/integration/test_ppo_pretraining_pipeline.py -q -k 'bc_pretraining_config_accepts_explicit_cpu_device or default_bc_pretraining_config_loads_auto_device or create_bc_trainer_threads_configured_device or pretraining_pipeline_smoke'` -> red before implementation: `4 failed`; failures were missing `device` field/argument/propagation.
  - `uv run ruff check --fix robot_sf/training/imitation_config.py scripts/training/pretrain_from_expert.py tests/integration/test_ppo_pretraining_pipeline.py && uv run ruff format robot_sf/training/imitation_config.py scripts/training/pretrain_from_expert.py tests/integration/test_ppo_pretraining_pipeline.py` -> passed.
  - `uv run pytest tests/integration/test_ppo_pretraining_pipeline.py -q` -> `10 passed`.
  - `BASE_REF=origin/main PYTEST_NUM_WORKERS=8 scripts/dev/pr_ready_check.sh` -> `3540 passed, 12 skipped, 3 warnings in 261.65s`; changed-file coverage reported `robot_sf/training/imitation_config.py: 91.3% [WARN]`, above the 80% minimum, and docstring TODO gate passed.
- Evidence that the change works here:
  - The targeted tests prove default `auto`, explicit YAML `cpu`, and propagation into both BC trainer construction points.
- Benchmarks or smoke tests, if applicable:
  - Not applicable; this changes BC training configuration plumbing, not benchmark runtime behavior.

## Risks / Rollout
- Compatibility risks: low; configs without `device` still default to `auto`.
- Failure modes: accelerator selection remains delegated to SB3/imitation when `auto` is used, so host-specific accelerator availability can still affect performance.
- Rollback or fallback plan: set `device: cpu` in BC YAMLs to force the previous conservative execution policy.

## Docs / Provenance
- Updated docs: `docs/imitation_learning_pipeline.md`, `docs/dev_guide.md`.
- Relevant design or provenance notes: issue #1213 decision states default `auto`.
- Any assumptions that need to be preserved: this only affects BC pre-training, not PPO fine-tuning or other training device policies.

## Follow-Up Issues
- Deferred work: none.
- Issues opened for follow-up: none.

## Reviewer Notes
- The full readiness gate produced only existing-style TensorFlow/pygame noise plus the changed-file coverage warning noted above.
- Ignored coverage output under `output/coverage` was inspected as disposable validation output; no durable artifact promotion needed.

## Release Notes

* **New Features**
  * Behavioral cloning pretraining now supports configurable device selection, defaulting to automatic accelerator detection when available.
  * CPU-only mode available for deterministic or resource-constrained training runs.

* **Documentation**
  * Updated training guides explaining device configuration options and pretraining behavior.

* **Tests**
  * Added test coverage for device configuration in the pretraining pipeline.

[Review Change Stack](https://app.coderabbit.ai/change-stack/ll7/robot_sf_ll7/pull/1223?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: robot_sf/training/imitation_config.py [hunk 1]
```diff
@@ -239,6 +241,7 @@ def from_raw(  # noqa: PLR0913
             random_seeds=ensure_seed_tuple(random_seeds),
             training_config_path=training_config_path.resolve() if training_config_path else None,
             scenario_config_path=scenario_config_path.resolve() if scenario_config_path else None,
+            device=str(device).strip() or "auto",
```
[the remark this answers is not in the document]
Comment (ll7) [reply robot_sf/training/imitation_config.py]: Fixed in a28bc0b7. `BCPretrainingConfig.from_raw(...)` now normalizes falsey/null device inputs to `"auto"` before string coercion, and tests/integration/test_ppo_pretraining_pipeline.py::test_bc_pretraining_config_treats_none_device_as_auto covers the YAML-null style case. Focused proof: `uv run --active pytest tests/integration/test_ppo_pretraining_pipeline.py -q -k "bc_pretraining_config_accepts_explicit_cpu_device or bc_pretraining_config_treats_none_device_as_auto or default_bc_pretraining_config_loads_auto_device or create_bc_trainer_threads_configured_device"`.

**Comment by ll7** [comment]
Autonomous draft review result: reviewed, not merge-ready only because this PR is still a draft.

Issue #1213 scope maps to the diff: BC pretraining now has a config-driven `device` field, default `auto`, explicit `cpu` override coverage, docs/config updates, and trainer propagation tests. Review thread is resolved and checks are green on head `a28bc0b7`.

No additional blocker found in this pass. Remove draft status before merge-ready can be considered.

**Comment by ll7** [comment]
Autonomous review pass found and fixed one remaining review-thread issue in commit `123dcebd`.

Fix:
- Changed `BCPretrainingConfig.from_raw(..., device=...)` from `str` to `str | None` so the public constructor type matches the existing `None` -> `auto` normalization behavior.

Validation after the fix:
- `./.venv/bin/python -m pytest tests/integration/test_ppo_pretraining_pipeline.py -q` -> 11 passed
- `./.venv/bin/ruff check robot_sf/training/imitation_config.py scripts/training/pretrain_from_expert.py tests/integration/test_ppo_pretraining_pipeline.py` -> passed
- `./.venv/bin/ruff format --check robot_sf/training/imitation_config.py scripts/training/pretrain_from_expert.py tests/integration/test_ppo_pretraining_pipeline.py` -> passed
- `git diff --check origin/main...HEAD` -> passed

I resolved the CodeRabbit review thread. Post-push GitHub checks are running on the new commit, so I am withholding `merge-ready` until those checks complete successfully.

**Comment by ll7** [comment]
Autonomous readiness review: `merge-ready`.

Proof basis:
- Closes #1213 with a scoped config-driven BC pretraining device policy.
- Diff is limited to BC config loading, trainer device propagation, checked-in BC YAML defaults, docs, and targeted integration tests.
- Review threads are resolved, including the `None` -> `auto` typing/normalization fix in `123dcebd`.
- GitHub checks on the current head are green: fast-feedback, smoke-artifacts, aggregate CI, promoted-planner-smoke, and CodeRabbit.
- Prior local proof after the fix: `tests/integration/test_ppo_pretraining_pipeline.py -q`, targeted ruff, format check, and `git diff --check origin/main...HEAD` passed.

No follow-up issue needed from this review pass.
